runtime/src/slots.rs

```diff
@@ -0,0 +1,1425 @@
+			#[compact] last_slot: LeasePeriodOf<T>,
+			#[compact] amount: BalanceOf<T>
+		) {
+			let who = ensure_signed(origin)?;
+			let bidder = Bidder::New(NewBidder{who: who.clone(), sub});
+			Self::handle_bid(bidder, auction_index, first_slot, last_slot, amount)?;
+		}
+
+		/// Make a new bid from a parachain account for renewing that (pre-existing) parachain.
+		///
+		/// The origin *must* be a parachain account.
+		///
+		/// Multiple simultaneous bids from the same bidder are allowed only as long as all active
+		/// bids overlap each other (i.e. are mutually exclusive). Bids cannot be redacted.
+		///
+		/// - `auction_index` is the index of the auction to bid on. Should just be the present
+		/// value of `AuctionCounter`.
+		/// - `first_slot` is the first lease period index of the range to bid on. This is the
+		/// absolute lease period index value, not an auction-specific offset.
+		/// - `last_slot` is the last lease period index of the range to bid on. This is the
+		/// absolute lease period index value, not an auction-specific offset.
+		/// - `amount` is the amount to bid to be held as deposit for the parachain should the
+		/// bid win. This amount is held throughout the range.
+		fn bid_renew(origin,
+			#[compact] auction_index: AuctionIndex,
+			#[compact] first_slot: LeasePeriodOf<T>,
+			#[compact] last_slot: LeasePeriodOf<T>,
+			#[compact] amount: BalanceOf<T>
+		) {
+			let who = ensure_signed(origin)?;
+			let para_id = <ParaIdOf<T>>::try_from_account(&who)
+				.ok_or("account is not a parachain")?;
+			let bidder = Bidder::Existing(para_id);
+			Self::handle_bid(bidder, auction_index, first_slot, last_slot, amount)?;
+		}
+
+		/// Set the off-boarding information for a parachain.
+		///
+		/// The origin *must* be a parachain account.
+		///
+		/// - `dest` is the destination account to receive the parachain's deposit.
+		fn set_offboarding(origin, dest: <T::Lookup as StaticLookup>::Source) {
+			let who = ensure_signed(origin)?;
+			let dest = T::Lookup::lookup(dest)?;
+			let para_id = <ParaIdOf<T>>::try_from_account(&who)
+				.ok_or("not a parachain origin")?;
+			<Offboarding<T>>::insert(para_id, dest);
+		}
+
+		/// Set the deploy information for a successful bid to deploy a new parachain.
+		///
+		/// This may be called before or after the beginning of the parachain's first lease period.
+		/// If called before then the parachain will become active at the first block of its
+		/// starting lease period. If after, then it will become active immediately after this call.
+		///
+		/// - `origin` must be the successful bidder account.
+		/// - `sub` is the sub-bidder ID of the bidder.
+		/// - `para_id` is the parachain ID allotted to the winning bidder.
+		/// - `code` is the parachain's Wasm validation function.
+		/// - `initial_head_data` is the parachain's initial head data.
+		fn set_deploy_data(origin,
+			#[compact] sub: SubId,
+			#[compact] para_id: ParaIdOf<T>,
+			code: Vec<u8>,
```

Comment:
Should we do some sort of size limit to these bytearrays? Seems like there is a potential to attack the system with huge blobs which ultimately get stored.

Reply:
there's an inherent limit by way of transaction/block size which i think is sufficient for now. these blobs could get legitimately huge and take up most of a block so i don't think a significantly lower limit makes sense. we may want to introduce a secure off-chain solution (e.g. some properly attested ipfs-based thing) for these massive code blobs eventually and only store the hash on-chain. but that's for post-1.0.